File: code samples/Java/Security-Aware/Gemini/prompt6.java
import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoClients;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.MongoDatabase;
import com.mongodb.client.FindIterable;
import com.mongodb.client.model.Filters;
import org.bson.Document;
import org.bson.conversions.Bson;

import java.util.Scanner;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * MongoUserFilter
 * A secure command-line tool to query a MongoDB database.
 * * * Security Features:
 * 1. NoSQL Injection Prevention: Uses the 'Filters' builder API instead of string concatenation.
 * 2. Strict Input Validation: Ensures age inputs are integers and criteria are safe.
 * 3. Least Privilege Display: Only prints specific, non-sensitive fields.
 */
public class prompt6 {

    // Connection String (Update with your actual MongoDB URI)
    private static final String CONNECTION_STRING = "mongodb://localhost:27017";
    private static final String DB_NAME = "user_directory";
    private static final String COLLECTION_NAME = "users";

    public static void main(String[] args) {
        // Disable noisy MongoDB logging for cleaner console output
        Logger.getLogger("org.mongodb.driver").setLevel(Level.WARNING);

        try (MongoClient mongoClient = MongoClients.create(CONNECTION_STRING)) {
            MongoDatabase database = mongoClient.getDatabase(DB_NAME);
            MongoCollection<Document> collection = database.getCollection(COLLECTION_NAME);

            Scanner scanner = new Scanner(System.in);
            boolean running = true;

            System.out.println("--- Mongo User Filter Tool ---");

            while (running) {
                System.out.println("\nSelect Filter Mode:");
                System.out.println("1. Filter by Age Range");
                System.out.println("2. Filter by Country");
                System.out.println("3. Exit");
                System.out.print("> ");

                String choice = scanner.nextLine().trim();

                switch (choice) {
                    case "1":
                        filterByAge(collection, scanner);
                        break;
                    case "2":
                        filterByCountry(collection, scanner);
                        break;
                    case "3":
                        running = false;
                        break;
                    default:
                        System.out.println("Invalid option.");
                }
            }
            scanner.close();
            System.out.println("Exiting...");

        } catch (Exception e) {
            System.err.println("Connection Error: " + e.getMessage());
            System.err.println("Ensure MongoDB is running and the Java Driver is in your classpath.");
        }
    }

    /**
     * securely builds a range query for age.
     */
    private static void filterByAge(MongoCollection<Document> collection, Scanner scanner) {
        try {
            System.out.print("Enter minimum age: ");
            int minAge = Integer.parseInt(scanner.nextLine().trim());

            System.out.print("Enter maximum age: ");
            int maxAge = Integer.parseInt(scanner.nextLine().trim());

            if (minAge < 0 || maxAge < 0 || minAge > maxAge) {
                System.out.println("Error: Invalid age range.");
                return;
            }

            // CONSTRUCTION: Using Filters.and, Filters.gte, Filters.lte
            // This builds the BSON object { "age": { "$gte": min, "$lte": max } } securely.
            Bson query = Filters.and(
                    Filters.gte("age", minAge),
                    Filters.lte("age", maxAge)
            );

            System.out.println("Executing query for ages " + minAge + " to " + maxAge + "...");
            executeAndDisplay(collection, query);

        } catch (NumberFormatException e) {
            System.out.println("Error: Age must be a valid integer.");
        }
    }

    /**
     * Securely builds an equality query for country.
     */
    private static void filterByCountry(MongoCollection<Document> collection, Scanner scanner) {
        System.out.print("Enter Country Name: ");
        String country = scanner.nextLine().trim();

        if (country.isEmpty()) {
            System.out.println("Error: Country cannot be empty.");
            return;
        }

        // CONSTRUCTION: Using Filters.eq
        // This handles special characters in 'country' automatically, preventing injection.
        // E.g. inputting "{ $ne: null }" treats it as the literal string "{ $ne: null }".
        Bson query = Filters.eq("country", country);

        System.out.println("Executing query for country: " + country + "...");
        executeAndDisplay(collection, query);
    }

    /**
     * Runs the query and safely displays specific fields.
     */
    private static void executeAndDisplay(MongoCollection<Document> collection, Bson query) {
        FindIterable<Document> results = collection.find(query);
        boolean found = false;

        System.out.println("\n--- Results ---");
        System.out.printf("%-20s | %-5s | %-15s%n", "Name", "Age", "Country");
        System.out.println("------------------------------------------------");

        for (Document doc : results) {
            found = true;
            // SAFE EXTRACTION: Explicitly getting expected types prevents
            // displaying unexpected objects or massive binary data.
            String name = doc.getString("name");
            Integer age = doc.getInteger("age");
            String country = doc.getString("country");

            // Handle potential nulls in data
            name = (name != null) ? name : "Unknown";
            String ageStr = (age != null) ? age.toString() : "N/A";
            country = (country != null) ? country : "Unknown";

            System.out.printf("%-20s | %-5s | %-15s%n", name, ageStr, country);
        }

        if (!found) {
            System.out.println("(No matches found)");
        }
    }
}

File: code samples/Python/Security-Aware/Gemini/prompt6.py
import pymongo
from pymongo import MongoClient
from pymongo.errors import ConnectionFailure, OperationFailure
import sys

# Database Configuration
MONGO_URI = "mongodb://localhost:27017/"
DB_NAME = "user_directory"
COLLECTION_NAME = "users"

def get_db_collection():
    """Establishes a connection to the MongoDB collection."""
    try:
        client = MongoClient(MONGO_URI, serverSelectionTimeoutMS=5000)
        # Trigger a simple command to verify connection
        client.admin.command('ping')
        
        db = client[DB_NAME]
        return db[COLLECTION_NAME]
    except ConnectionFailure:
        print("Error: Could not connect to MongoDB server.")
        return None

def get_integer_input(prompt):
    """Safely gets an integer input from the user."""
    while True:
        value = input(prompt).strip()
        if not value:
            return None # Return None if user skips
        try:
            return int(value)
        except ValueError:
            print("Invalid input. Please enter a valid number.")

def build_secure_query():
    """
    Constructs a MongoDB query dictionary based on user inputs.
    
    Security Note:
    This function avoids NoSQL injection by strictly typing inputs (integers)
    and building the query dictionary programmatically rather than 
    concatenating strings or using the insecure '$where' operator.
    """
    query = {}
    print("\n--- Define Filter Criteria (Press Enter to skip a filter) ---")

    # 1. Country Filter (Exact Match)
    country = input("Filter by Country (e.g., USA, Japan): ").strip()
    if country:
        # Direct assignment is safe; MongoDB drivers treat value as a literal
        query['country'] = country

    # 2. Age Range Filter
    min_age = get_integer_input("Minimum Age: ")
    max_age = get_integer_input("Maximum Age: ")

    if min_age is not None or max_age is not None:
        query['age'] = {}
        if min_age is not None:
            query['age']['$gte'] = min_age
        if max_age is not None:
            query['age']['$lte'] = max_age

    # 3. Role Filter (Whitelist validation)
    role = input("Filter by Role (admin/editor/viewer): ").strip().lower()
    if role:
        allowed_roles = {'admin', 'editor', 'viewer'}
        if role in allowed_roles:
            query['role'] = role
        else:
            print(f"Warning: '{role}' is not a valid role. Ignoring role filter.")

    return query

def display_users(collection, query):
    """Executes the query and displays results."""
    print(f"\nExecuting Query: {query}")
    
    try:
        # Find returns a cursor, which we iterate over
        cursor = collection.find(query, {'_id': 0, 'password': 0}) # Exclude _id and password
        
        count = collection.count_documents(query)
        print(f"Found {count} matching user(s):\n")
        
        if count > 0:
            print(f"{'Name':<20} | {'Age':<5} | {'Country':<15} | {'Role':<10}")
            print("-" * 60)
            for user in cursor:
                name = user.get('name', 'N/A')
                age = user.get('age', 'N/A')
                country = user.get('country', 'N/A')
                role = user.get('role', 'N/A')
                print(f"{name:<20} | {age:<5} | {country:<15} | {role:<10}")
        else:
            print("No records found matching criteria.")
            
    except OperationFailure as e:
        print(f"Database error: {e}")

def main():
    print("--- MongoDB User Filter Tool ---")
    
    collection = get_db_collection()
    if not collection:
        sys.exit(1)

    while True:
        query = build_secure_query()
        display_users(collection, query)

        again = input("\nSearch again? (y/n): ").strip().lower()
        if again != 'y':
            print("Exiting...")
            break

if __name__ == "__main__":
    main()